### PR TITLE
Detokenizer fixes

### PR DESCRIPTION
**Pull request by jaime-m-p**
This PR tries to solve most common problems with detokenization:
 - spaces after special tokens
 - spaces before and after apostrophes

Related issues: https://github.com/ggerganov/llama.cpp/issues/8023, https://github.com/ggerganov/llama.cpp/issues/7938.

___
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [ ] Low
  - [x] Medium
  - [ ] High

### CODE REVIEW

**Comment by cmp-nct** [comment]
I gave it a test run using phi3-mini-instruct
First I ran a string through the tokenize_test.py which the model came with:
Both of those two strings tokenize identical (the \n is ignore for some reason)
`Test:<|user|>This is a test<|end|>\n<|assistant|>`  
`Test:<|user|>This is a test<|end|><|assistant|>`
```
ID: 1, Content: '<s>'
ID: 4321, Content: '▁Test'
ID: 29901, Content: ':'
ID: 32010, Content: '<|user|>'
ID: 910, Content: '▁This'
ID: 338, Content: '▁is'
ID: 263, Content: '▁a'
ID: 1243, Content: '▁test'
ID: 32007, Content: '<|end|>'
ID: 32001, Content: '<|assistant|>'
```

Now I ran the same through llama.cpp tokenization:
```
<|assistant|>':
  4321 -> ' Test'
 29901 -> ':'
 32010 -> '<|user|>'
   910 -> ' This'
   338 -> ' is'
   263 -> ' a'
  1243 -> ' test'
 32007 -> '<|end|>'
 29871 -> ' '
    13 -> '
'
 32001 -> '<|assistant|>'
```

It looks like the python tokenizer does not add anything between two special tokens.

**Comment by cmp-nct** [comment]
@jaime-m-p 
Oh, I didn't know about that behavior. Explains a lot!
1) The \n was manually added by me based on the official template by Microsoft
image
But as you pointed out: END has "rstrip" true, so the newline is removed.

2) model name special behavior
That's great to know. It also explains why it's not working for me!
I am testing this on a recent xtuner phi3-mini model (llava type) which comes with genetic.name == "LLaMA v2"

I'll repeat the tests after fixing those issues and reverting my changes. Given your results that's promising.

It's a little troublesome that such errors can very easily sneak into a model and it's very hard to notice them, even harder to fix them without blindly recreating the model from originals.
Likely the majority of GGUF files on huggingspace are flawed in some way.

**Comment by Harsha-Nori** [comment]
Hi @jaime-m-p and @cmp-nct, really grateful you guys are looking into this! I'm traveling without reliable access to a computer at the moment, but wanted to ask if these fixes now keep stability on retokenization with Phi-3 (i.e. the roundtrip of text -> tokens -> text -> tokens results in the same tokens). The constant whitespace insertion on each cycle was causing serious kv-cache reuse issues on our side and I'm really hopeful that this update resolves it!

**Comment by Harsha-Nori** [comment]
Hmm, great point. I think what I'm really hoping for is eventual stability on the second or third tokenize/detokenize cycles -- before your PR, Phi-3 had the problem of constantly changing the token_id at index 1 (due to growing spaces), which really caused issues.

I think this set of changes is good enough to solve most of our problems :).

File: tests/test-tokenizer-1-bpe.cpp [hunk 1]
```diff
@@ -136,6 +137,10 @@ int main(int argc, char **argv) {
         for (auto & t : threads) {
             t.join();
         }
+
+        if(errcode) {
```
Comment (ggerganov) [inline tests/test-tokenizer-1-bpe.cpp:141]:
```suggestion
        if (errcode) {
```

**Review by ggerganov — APPROVED** [review]
Nice work

**Pull request merged**